source_code/main_mcu/src/LOGIC/logic_totp.c

```diff
@@ -0,0 +1,174 @@
+/*
+ * This file is part of the Mooltipass Project (https://github.com/mooltipass).
+ * Copyright (c) 2019 Stephan Mathieu
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+/*!  \file     logic_totp.c
+*    \brief    General logic for TOTP algorithm based on RFC 6238 (https://tools.ietf.org/html/rfc6238)
+*    Created:  31/07/2020
+*    Author:   dnns01
+*/
+#include "logic_totp.h"
+
+/*! \fn     logic_totp_sha1_hmac(void *out, const void *key, const void *message)
+*   \brief  Calculate SHA1 HMAC Hash for given message using given key.
+*   \param  calculated hash, char[20] expected
+*   \param  key for SHA1 HMAC calculation, char[10] expected
+*   \param  message to be hashed, 8 bytes expected
+*   \return void
+*/
+static void logic_totp_sha1_hmac(void* out, const void *key, const void *message)
+{
+    br_hmac_key_context key_context;
+    br_hmac_context context;
+
+    br_hmac_key_init(&key_context, &br_sha1_vtable, key, LOGIC_TOTP_SHA1_HMAC_KEY_LEN);
+    br_hmac_init(&context, &key_context, 0);
+    br_hmac_update(&context, message, LOGIC_TOTP_SHA1_HMAC_MESSAGE_LEN);
+    br_hmac_out(&context, out);
+}
+
+
+/*! \fn     logic_totp_get_time()
+*   \brief  Retrieves time from RTC and converts calendar back to unix timestamp
+*   \return current unix timestamp
+*/
+static time_t logic_totp_get_time(void)
+{
+    calendar_t calendar;
+    struct tm datetime;
+
+    timer_get_calendar(&calendar);
+
+    datetime.tm_sec = calendar.bit.SECOND;
+    datetime.tm_min = calendar.bit.MINUTE;
+    datetime.tm_hour = calendar.bit.HOUR;
+    datetime.tm_mday = calendar.bit.DAY;
+    datetime.tm_mon = calendar.bit.MONTH - 1;
+    datetime.tm_year = calendar.bit.YEAR + 2000 - 1900;
+
+    // Assumption: If running in the emulator, gmtime is used. Therefore we call _mkgmtime to get the current timestamp, whereas
+    // the real device is running on UTC, therefore mktime is used to get the timestamp.
+#ifndef EMULATOR_BUILD
+    return mktime(&datetime);
```

Comment:
you're likely to suffer from https://en.wikipedia.org/wiki/Year_2038_problem 
As with my other comment, I **strongly** recommend you to make / copy your own mktime that uses an uint64_t :)

Reply:
Ok, what I just figured out, the emulator (at least on my computer) has a 64 bit long time_t. And TOTP wants to have 8 bytes as message to be hashed. Therefore, not only regarding Year 2038 problem, we need 64 bit. Correct me if I am wrong, but in order to achieve that, the calculation from calendar_t to uint64_t has to be done on our own. But that means, in any case, the time has to be either in a fixed timezone, or there has to be an indicator, which timezone it is. (And if the time is not in a fixed timezone for every device, there has to be an indicator for DST as well.)

Reply:
I was actually going to suggest to limpkin that the time that is set on the device (with set_calendar) is UTC and then we would have to have a way to set the timezone.

Here is my implementation of time() to get UNIX time. You are free to use this if you want or I can create a pull request for limpkin for this code:

`
#define IS_LEAP_YEAR(year)  ((((year) % 4 == 0) && ((year) % 100 != 0)) || ((year) % 400 == 0))
#define SEC_IN_HOUR (60 * 60)
#define SEC_IN_DAY (24 * SEC_IN_HOUR)

/* Constants */
//Days Per Month
static uint8_t const dph[] = { 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 };
static uint8_t const LEAP_MONTH = 1; //February
static uint16_t const EPOCH_YEAR = 1970;

/*!	\fn		time(void)
*	\brief	Get the current unix time
*   \param  None
*   \return Seconds since 1970
*/
uint64_t time(void)
{
    uint64_t unix_time = 0;
    int32_t i;
    uint32_t num_days = 0;
    calendar_t tm;
    uint8_t is_curr_year_leap;
    uint16_t year;

    timer_get_calendar(&tm);
    year = tm.bit.YEAR + 2000;
    is_curr_year_leap = IS_LEAP_YEAR(year);

    //Add up days in all years
    for (i = EPOCH_YEAR; i < year; ++i)
    {
        num_days += IS_LEAP_YEAR(i) ? 366 : 365;
    }

    //Add up days in all months
    for (i = 0; i < tm.bit.MONTH - 1; ++i)
    {
        num_days += dph[i];
        num_days += ((i == LEAP_MONTH) && is_curr_year_leap) ? 1 : 0;
    }

    //Add days in month
    num_days += tm.bit.DAY - 1;

    //Add the days hours, minutes, and seconds
    unix_time = num_days * SEC_IN_DAY;
    unix_time += tm.bit.HOUR * SEC_IN_HOUR;
    unix_time += tm.bit.MINUTE * 60;
    unix_time += tm.bit.SECOND;

    return unix_time;
}
`